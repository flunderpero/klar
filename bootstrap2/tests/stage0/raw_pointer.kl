struct Person:
    name str
    age i32
end

fn get_person(buf ptr, index i32) Person:
    let p = getptr(buf, index)
    return p
end

fn main():
    mut buf = malloc(16)
    setptr(buf, "Hi", 0)
    setptr(buf, "there", 1)
    print(getptr(buf, 0))
    print(getptr(buf, 1))
    buf = realloc(buf, 32)
    print(getptr(buf, 0))
    print(getptr(buf, 1))
    let p = Person("Bob", 42)
    setptr(buf, p, 2)
    let p2 = get_person(buf, 2)
    print(f"{p2.name} is {p2.age} years old")
end
