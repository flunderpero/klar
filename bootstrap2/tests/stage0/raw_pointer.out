Hi
there
Hi
there
Bob is 42 years old
